aperturedb/onion.go

```diff
@@ -0,0 +1,163 @@
+package aperturedb
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/lightninglabs/aperture/aperturedb/sqlc"
+	"github.com/lightningnetwork/lnd/clock"
+	"github.com/lightningnetwork/lnd/tor"
+)
+
+type (
+	NewOnionPrivateKey = sqlc.InsertOnionParams
+)
+
+// OnionDB is an interface that defines the set of operations that can be
+// executed against the onion database.
+type OnionDB interface {
+	// InsertOnion inserts a new onion private key into the database.
+	InsertOnion(ctx context.Context, arg NewOnionPrivateKey) error
+
+	// SelectOnionPrivateKey selects the onion private key from the
+	// database.
+	SelectOnionPrivateKey(ctx context.Context) ([]byte, error)
+
+	// DeleteOnionPrivateKey deletes the onion private key from the
+	// database.
+	DeleteOnionPrivateKey(ctx context.Context) error
+}
+
+// OnionTxOptions defines the set of db txn options the OnionStore
+// understands.
+type OnionDBTxOptions struct {
+	// readOnly governs if a read only transaction is needed or not.
+	readOnly bool
+}
+
+// ReadOnly returns true if the transaction should be read only.
+//
+// NOTE: This implements the TxOptions
+func (a *OnionDBTxOptions) ReadOnly() bool {
+	return a.readOnly
+}
+
+// NewOnionDBReadTx creates a new read transaction option set.
+func NewOnionDBReadTx() OnionDBTxOptions {
+	return OnionDBTxOptions{
+		readOnly: true,
+	}
+}
+
+// BatchedOnionDB is a version of the OnionDB that's capable of batched
+// database operations.
+type BatchedOnionDB interface {
+	OnionDB
+
+	BatchedTx[OnionDB]
+}
+
+// OnionStore represents a storage backend.
+type OnionStore struct {
+	db    BatchedOnionDB
+	clock clock.Clock
+}
+
+// NewOnionStore creates a new OnionStore instance given a open BatchedOnionDB
+// storage backend.
+func NewOnionStore(db BatchedOnionDB) *OnionStore {
+	return &OnionStore{
+		db:    db,
+		clock: clock.NewDefaultClock(),
+	}
+}
+
+// StorePrivateKey stores the private key according to the implementation of
+// the OnionStore interface.
+func (o *OnionStore) StorePrivateKey(privateKey []byte) error {
+	ctxt, cancel := context.WithTimeout(
+		context.Background(), DefaultStoreTimeout,
+	)
+	defer cancel()
+
+	var writeTxOpts OnionDBTxOptions
+	err := o.db.ExecTx(ctxt, &writeTxOpts, func(tx OnionDB) error {
+		// Only store the private key if it doesn't already exist.
```

Comment:
Can be taken care of my making insert a no-op uspert. So would add that unique constraint, then catch the on conflict.

Reply:
updated :+1: